### ISSUE TITLE
[css-mediaqueries] Inverted-colors media feature can cause problems on semi-transparent images.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#9674]
https://drafts.csswg.org/mediaqueries-5/#inverted says:

> User agents must add the following rule to their [UA style sheet](https://drafts.csswg.org/css-cascade-5/#cascade-origin-ua):
> ```css
> @media (inverted-colors) {
>   img:not(picture>img), picture, video { filter: invert(100%); }
> }
> ```

However, this causes problems for icons and similar things that are meant to be semi-transparent above the background. Also, doesn't account for other things like background-image.

Consider [this straight-forward test-case](http://crisal.io/tmp/inverted-colors-icon.html). Per spec, when inverting colors, the icon is completely invisible.

Should we remove this rule, leaving it up to sites to invert only the images they want to invert? Or should we try to do something smarter?

cc @CanadaHonk @frivoal

**Comment by CanadaHonk** [comment]
This rule has caused some other minor issues for Gecko too. Also iirc Chromium did not add this UA rule when implementing the media feature (at all).

I like the idea behind it but I'm not sure it can be executed well without causing issues like these?

**Comment by tsbertalan** [comment]
On [bugzilla, I posted](url), I posted some related info, and was requested to comment here also. The gist is that Firefox (in the current version) tries to uninvert photographs after the OS has inverted the whole screen (e.g. Windows 10's built in [color filters](https://support.microsoft.com/en-us/windows/use-color-filters-in-windows-43893e44-b8b3-2e27-1a29-b0c15ef0e5ce) feature). However, aside from some bugs in that implementation leaving photos inverted after the user turns off the OS inversion (that bug and siblings), such smart logic should not be applied to images that are meant to be not naturalistic photos, but UI elements or text.

Such is a broader category than semi-transparent images, but the particular place I'm seeing the issue is with formulae on Wikipedia, which will blend in with the page background (which is black to my eyes, but still captured as white by screenshots). It can be fixed temporarily by adding

```css
.mw-invert {
  filter: invert(0);
}
```
in in a new style sheet in the "Style Editor" tab of the devtools for the page (I add this as a [Stylus](https://addons.mozilla.org/en-US/firefox/addon/styl-us/) profile).

In my opinion, the browser shouldn't guess at which images are "naturalistic" and thus shouldn't be inverted--either it gives the developers information on whether the OS is inverted and lets them do their own `filter:invert` reactively, or (if this is deemed insecure or something), allow devs to specify some extra naturalistic-image property to tag images that should not invert (which would mean a change of standard, so also probably no-go). Since both are bad for their own reasons, I think the browser should just do the most straightforward, user-predictable thing and not second-guess the OS's inversion.

**Comment by meyerweb** [comment]
> In my opinion, the browser shouldn't guess at which images are "naturalistic" and thus shouldn't be inverted… Since both are bad for their own reasons, I think the browser should just do the most straightforward, user-predictable thing and not second-guess the OS's inversion.

I agree.  This does mean extra work for authors, but it’s work they should be doing anyway — and in many cases, will _have_ to do even if the browser tries to guess which elements should be un-inverted, because it will inevitably guess wrong at some point.

**Comment by lukewarlow** [comment]
Is there a potential compat issue with removing this as Apple has shipped with it for ages?

**Comment by emilio** [comment]
My guess is not much, since it's a rather rare feature. We actually found compat issues from _shipping_ this.

cc @nt1m @smfr tho

**Comment by lukewarlow** [comment]
> Also iirc Chromium did not add this UA rule when implementing the media feature (at all).


Yeah Chrome doesn't support media feature queries in ua styles so the implementation is blocked by this rule. So from that perspective I'd love to not need this.

As mentioned above I do worry about compat. Has anyone from WebKit been asked about this?

**Comment by cookiecrook** [comment]
Removing this would break support for a major accessibility feature on multiple platforms.

>> Is there a potential compat issue with removing this as Apple has shipped with it for ages?

> My guess is not much, since it's a rather rare feature. We actually found compat issues from shipping this.

Not sure how you came to that conclusion... Content images (think every News site; most sites for that matter) represent the vast majority of images on the Web... and low-vision, light-sensitive users have told us over-and-over that they _need_ the Smart Invert feature for sites that don't support dark mode.

Removing this would break Smart Invert on most websites, to account for an uncommon edge case with a transparent SVG? This type of example is the primary place where the media query should be used by the web author... In the priorities of constituency context, removing this would penalize both all SI users and the vast majority of web authors.

**Comment by cookiecrook** [comment]
@meyerweb wrote:

> [removing default support for smart invert] does mean extra work for authors, but it’s work they should be doing anyway — and in many cases, will have to do even if the browser tries to guess which elements should be un-inverted, because it will inevitably guess wrong at some point.

Why make millions of developers do extra work in order to serve the vast majority of users of this feature? Most authors will not, penalizing the users.

- The defaults here get it right pf the vast majority of `<img>` usage on the web, without intervention needed by users or authors. 
- Authors can override it using the media feature in cases like the edge case listed above.
- Users (on Apple platforms, an maybe more) can switch from Smart Invert to Classic Invert, which results in identical GPU inversion, but without any API for native or web apps to know it's "on."

**Comment by CanadaHonk** [comment]
I suppose WebKit continues to agree with spec even with the PR merged. It doesn't say engines must not have the rule, rather engines do not have to have this just to implement the media query itself, so it could be an engine decision whether to do it or not? (Or if someone really wanted a pref somewhere too?)

**Comment by nt1m** [comment]
@CanadaHonk The current wording doesn't really allow for UAs to add a rule. It probably should at least allow that: "User agents may add ...". But based on what James said, I think it would be best to make this rule mandatory, if users don't like it they can always disable "smart invert" on the relevant platforms.

**Comment by lukewarlow** [comment]
Fwiw I agree that it's a beneficial default style but I also do wonder if

 the fact authors probably need to do changes to make this work fully (handling shadows for example), combined with the fact chrome can't add media feature queries to their UA styles without a change that based on my (admittedly limited) understanding would be non-trivial

make it acceptable for this to be up to the UA?

**Comment by emilio** [comment]
> Removing this would break Smart Invert on most websites, to account for an uncommon edge case with a transparent SVG? This type of example is the primary place where the media query should be used by the web author... In the priorities of constituency context, removing this would penalize both all SI users and the vast majority of web authors.

I don't think it's quite "an uncommon edge case with a transparent SVG". Contentful images are not uncommon. For example, this breaks all equations in wikipedia.

If you use smart invert and go to https://en.wikipedia.org/wiki/Equation#Description in Safari (or Firefox release for that matter), you'll just see black space where the equations and symbols are supposed to be.

Sure, the site _can_ fix it, and Wikipedia probably would if prompted to do so, but I think the failure mode (making the content completely invisible, no signal there was anything there), plus the fact that the site isn't really doing anything wrong, it's just the UA messing up the page, doesn't seem quite the right trade-off to me, even if the rule gets things right in 90% of the cases.

**Comment by emilio** [comment]
![image](https://github.com/w3c/csswg-drafts/assets/1323194/14993f9a-58b3-4b9f-8aa8-b4b1028af90f)

^ Is an example of the above in Wikipedia (left is Firefox Nightly, right Safari). It's un-inverted because macOS' screenshot tool somehow didn't account for the color inversion.

After a bit more casual browsing in Safari I also noticed this causes very low contrast on a lot of other pages (the icons in https://www.mathjax.org for example are invisible or almost invisible), so are the icons in Safari's developer tools.

**Comment by nt1m** [comment]
I wonder if the solution here is a special rule for SVG rather than removing the smart invert rule altogether.

**Comment by nt1m** [comment]
Something like:
```
@media (inverted-colors) {
  img:not(picture>img):not([src*="svg"]), picture, video { filter: invert(100%); }
}
```

**Comment by emilio** [comment]
Semi transparent PNG's are also affected. Also your rule wouldn't cover stuff like srcset, or `<source>`...

In general what you'd want is to only un-invert opaque images. One nice thing of doing it via CSS is that authors can override it if needed, but I'd tend to think that, would I be the one to implement this, maybe pushing it a bit further down and just make it magic / use the engine's knowledge about whether the image is opaque to decide whether to un-invert would be better here?

**Comment by cookiecrook** [comment]
> In general what you'd want is to only un-invert opaque images. 

brainstorming… new pseudo-class selector? 

```css
img:opaque {}
img:not(:transparent) {}
```

**Comment by emilio** [comment]
Maybe? It wouldn't really need to be exposed outside of user agent stylesheets. But that exposes one more bit from the (potentially cross-origin) image data to the embedder page, which might or might not be fine (@annevk had opinions on that iirc?)

**Comment by annevk** [comment]
Yeah, that seems bad. Perhaps we can make it a property instead? `auto-invert: on | off` or some such? That gives web developers control, without revealing when it happens.

**Comment by frivoal** [comment]
Reopening, on account of https://github.com/w3c/csswg-drafts/issues/9674#issuecomment-1901037121 and subsequent discussion.

**Comment by cookiecrook** [comment]
@emilio gave some good examples of bugs that I believe the Wikipedia and MathJax site owners should be resolved, either by the site owner or an update to the query. I've filed the WebKit one here: https://webkit.org/b/271959. 

I also want to illustrate what would be taken away from users if this inversion feature were removed from the UA style sheet.

Sites like the New York Times (which has never adopted dark mode, btw) would go from this:
![dark web page with images colored as expected; skin tones are normal; food is recognizable as food](https://github.com/w3c/csswg-drafts/assets/75867/639cccf3-aa9b-4b18-b342-4df3f0f11830)

Back to this:
![dark web page with inverted images; skin tones are blue; food is unrecognizable as food](https://github.com/w3c/csswg-drafts/assets/75867/a501eb59-69ed-4f3b-ac99-2905a43821ab)

Users have grown to rely on this "Smart Invert" feature for a long time in the Web... It's been shipping for 8 years (well before `prefers-color-scheme`) and works desirably in most web pages. Let's fix the newly reported issue of these transparent SVGs and PNGs without breaking the feature en masse.

**Comment by cookiecrook** [comment]
@emilio gave some good examples of bugs that should be resolved, either by the Wikipedia and MathJax site owners, or with an update to the UA style sheet selector. I've filed the WebKit one here: https://webkit.org/b/271959

I'd also like to give an example of what would be taken away from users if this inversion feature were removed. I believe this type of image is much more common than the transparent icons and formulas mentioned above.

Sites like the New York Times (which has never adopted dark mode, btw) would go from this (as shipping today):
![dark web page with expected photo colors; skin tones normal; food looks like food](https://github.com/w3c/csswg-drafts/assets/75867/28d008bb-db92-41bc-bb31-55a176ffb96d)

Back to this:
![dark web page with unexpectedly inverted photo colors; skin tones are blue; food is unrecognizable as food](https://github.com/w3c/csswg-drafts/assets/75867/a46f1d26-aa23-4953-bc68-669774836913)

Users have come to rely on smart inversion for a long time on the web. It's been shipping for 8 years (long before `prefers-color-scheme`) and works desirably and well for most web pages. Let's fix the newly reported issue with transparent PNGs and SVGs, but not gut the feature en masse.
Smart Invert has been shipping for 8 years, since

**Comment by emilio** [comment]
For the record, we removed this and we haven't seen any regression report so far. We did get reports about the missing mathjax formulas on Wikipedia, tho.

**Comment by lukewarlow** [comment]
I guess one thing to say is it might not be considered a regression by users as Firefox haven't shipped (at least until very recently?) that media query and UA rule, whereas WebKit has for years.

**Comment by SebastianZ** [comment]
The spec. now reads like this:

> inverted
> All pixels within the displayed area have been inverted.
> 
> This value must not match if the user agent has done some kind of content aware inversion such as one that preserves the images.

That means authors have no way to detect whether UAs use smart inversion and take counter measures for mistakes.

@annevk wrote:

> Yeah, that seems bad. Perhaps we can make it a property instead? `auto-invert: on | off` or some such? That gives web developers control, without revealing when it happens.

I assume that's meant as feature in addition to the media feature, right?

-----

It seems to me we have two options here.

1. Disallow UAs to do smart inversion

   This makes color inversion predictable but puts the responsibility of correcting incorrectly converted parts completely on authors.

2. Continue allowing UAs to do smart inversion

   This requires a way for authors to correct parts that got incorrectly handled by the UA.
   In this case we have four cases:
   1. Something got correctly inverted => good
   2. Something correctly stayed uninverted => good
   3. Something got incorrectly inverted
   4. Something incorrectly stayed uninverted

   The third case could be covered by a property as suggested by @annevk.
   Question is what to do in the fourth case. I see two solutions to this. Either also cover this via an additional value for the suggested property that forces inversion. Then I'd rather call this property `color-inversion` with the values `auto`, `none`, and `force`. Or it could be covered by a pseudo-class like `:color-inverted` (and target all those _not_ auto-inverted via `:not(:color-inverted)`.

Sebastian
